### ISSUE TITLE
Wrong DDL generated when using precision on TIMESTAMPTZ type

### ISSUE THREAD

**Issue by zero88** [jOOQ/jOOQ#8965]
### Expected behavior and actual behavior:

Currently, `Jooq` generate SQL to execute field with data type: `time/timestamp with time zone` as

`"F_TIMESTAMPZ" timestamp with time zone(6)`

But in `H2/PostgreSQL/Oracle` the definition must be

`"F_TIMESTAMPZ" timestamp(<precision>) with time zone`

Refs: [Oracle](https://docs.oracle.com/cd/B19306_01/server.102/b14225/ch4datetime.htm#i1006081), [PostgreSQL](https://www.postgresql.org/docs/10/datatype-datetime.html), [H2](https://h2database.com/html/datatypes.html#timestamp_with_time_zone_type)

- Inspect code
https://github.com/jOOQ/jOOQ/blob/master/jOOQ/src/main/java/org/jooq/impl/Tools.java#L4762-L4767
https://github.com/jOOQ/jOOQ/blob/master/jOOQ/src/main/java/org/jooq/impl/SQLDataType.java#L456
https://github.com/jOOQ/jOOQ/blob/master/jOOQ/src/main/java/org/jooq/impl/SQLDataType.java#L490

- **Log**

```bash
09:43:18.342 [vert.x-eventloop-thread-0] DEBUG org.jooq.tools.LoggerListener - Exception 
org.jooq.exception.DataAccessException: SQL [create table if not exists "mock1"."TBL_SAMPLE_01"("ID" int not null auto_increment, "F_BOOL" boolean null, "F_DATE" timestamp(6) null, "F_STR" varchar(31) null, "F_VALUE_JSON" clob null, "F_TIMESTAMP" timestamp(6) null, "F_TIMESTAMPZ" timestamp with time zone(6) null, "F_DATE_1" date null, "F_TIME" time(8) null, "F_DURATION" varchar(20) null, "F_PERIOD" varchar(20) null)]; Syntax error in SQL statement "CREATE TABLE IF NOT EXISTS ""mock1"".""TBL_SAMPLE_01""(""ID"" INT NOT NULL AUTO_INCREMENT, ""F_BOOL"" BOOLEAN NULL, ""F_DATE"" TIMESTAMP(6) NULL, ""F_STR"" VARCHAR(31) NULL, ""F_VALUE_JSON"" CLOB NULL, ""F_TIMESTAMP"" TIMESTAMP(6) NULL, ""F_TIMESTAMPZ"" TIMESTAMP WITH TIME ZONE([*]6) NULL, ""F_DATE_1"" DATE NULL, ""F_TIME"" TIME(8) NULL, ""F_DURATION"" VARCHAR(20) NULL, ""F_PERIOD"" VARCHAR(20) NULL) "; expected "FOR, UNSIGNED, INVISIBLE, VISIBLE, NOT, NULL, AS, DEFAULT, GENERATED, ON, NOT, NULL, AUTO_INCREMENT, BIGSERIAL, SERIAL, IDENTITY, NULL_TO_DEFAULT, SEQUENCE, SELECTIVITY, COMMENT, CONSTRAINT, PRIMARY, UNIQUE, NOT, NULL, CHECK, REFERENCES, ,, )"; SQL statement:
create table if not exists "mock1"."TBL_SAMPLE_01"("ID" int not null auto_increment, "F_BOOL" boolean null, "F_DATE" timestamp(6) null, "F_STR" varchar(31) null, "F_VALUE_JSON" clob null, "F_TIMESTAMP" timestamp(6) null, "F_TIMESTAMPZ" timestamp with time zone(6) null, "F_DATE_1" date null, "F_TIME" time(8) null, "F_DURATION" varchar(20) null, "F_PERIOD" varchar(20) null) [42001-197]
	at org.jooq_3.11.11.H2.debug(Unknown Source)
	at org.jooq.impl.Tools.translate(Tools.java:2430)
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:832)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:364)
	at mypackage.SQLHelper.createTable(SQLHelper.java:163)
	at mypackage.SQLHelper.createTableAndIndex(SQLHelper.java:149)
	at mypackage.SQLHelper.lambda$createNewDatabase$4(SQLHelper.java:120)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:270)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at mypackage.SQLHelper.createNewDatabase(SQLHelper.java:124)
	at mypackage.SQLHelper.createDatabase(SQLHelper.java:102)
	at mypackage.SQLHelper.start(SQLHelper.java:71)
	at io.vertx.core.impl.DeploymentManager.lambda$doDeploy$8(DeploymentManager.java:494)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
	at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:462)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:897)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.h2.jdbc.JdbcSQLException: Syntax error in SQL statement "CREATE TABLE IF NOT EXISTS ""mock1"".""TBL_SAMPLE_01""(""ID"" INT NOT NULL AUTO_INCREMENT, ""F_BOOL"" BOOLEAN NULL, ""F_DATE"" TIMESTAMP(6) NULL, ""F_STR"" VARCHAR(31) NULL, ""F_VALUE_JSON"" CLOB NULL, ""F_TIMESTAMP"" TIMESTAMP(6) NULL, ""F_TIMESTAMPZ"" TIMESTAMP WITH TIME ZONE([*]6) NULL, ""F_DATE_1"" DATE NULL, ""F_TIME"" TIME(8) NULL, ""F_DURATION"" VARCHAR(20) NULL, ""F_PERIOD"" VARCHAR(20) NULL) "; expected "FOR, UNSIGNED, INVISIBLE, VISIBLE, NOT, NULL, AS, DEFAULT, GENERATED, ON, NOT, NULL, AUTO_INCREMENT, BIGSERIAL, SERIAL, IDENTITY, NULL_TO_DEFAULT, SEQUENCE, SELECTIVITY, COMMENT, CONSTRAINT, PRIMARY, UNIQUE, NOT, NULL, CHECK, REFERENCES, ,, )"; SQL statement:
create table if not exists "mock1"."TBL_SAMPLE_01"("ID" int not null auto_increment, "F_BOOL" boolean null, "F_DATE" timestamp(6) null, "F_STR" varchar(31) null, "F_VALUE_JSON" clob null, "F_TIMESTAMP" timestamp(6) null, "F_TIMESTAMPZ" timestamp with time zone(6) null, "F_DATE_1" date null, "F_TIME" time(8) null, "F_DURATION" varchar(20) null, "F_PERIOD" varchar(20) null) [42001-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357)
	at org.h2.message.DbException.getSyntaxError(DbException.java:217)
	at org.h2.command.Parser.getSyntaxError(Parser.java:555)
	at org.h2.command.Parser.read(Parser.java:3518)
	at org.h2.command.Parser.readIfMore(Parser.java:950)
	at org.h2.command.Parser.parseCreateTable(Parser.java:6571)
	at org.h2.command.Parser.parseCreate(Parser.java:4615)
	at org.h2.command.Parser.parsePrepared(Parser.java:380)
	at org.h2.command.Parser.parse(Parser.java:335)
	at org.h2.command.Parser.parse(Parser.java:311)
	at org.h2.command.Parser.prepareCommand(Parser.java:278)
	at org.h2.engine.Session.prepareLocal(Session.java:611)
	at org.h2.engine.Session.prepareCommand(Session.java:549)
	at org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1247)
	at org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)
	at org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:304)
	at com.zaxxer.hikari.pool.ProxyConnection.prepareStatement(ProxyConnection.java:311)
	at com.zaxxer.hikari.pool.HikariProxyConnection.prepareStatement(HikariProxyConnection.java)
	at org.jooq.impl.ProviderEnabledConnection.prepareStatement(ProviderEnabledConnection.java:109)
	at org.jooq.impl.SettingsEnabledConnection.prepareStatement(SettingsEnabledConnection.java:73)
	at org.jooq.impl.AbstractQuery.prepare(AbstractQuery.java:416)
	at org.jooq.impl.AbstractQuery.execute(AbstractQuery.java:322)
	... 28 common frames omitted
```
 
- `DDL`
```sql
CREATE SCHEMA IF NOT EXISTS "mock1";
SET SCHEMA "mock1";
CREATE TABLE IF NOT EXISTS tbl_sample_01 (
    id INT AUTO_INCREMENT,
    f_bool bit,
    f_date TIMESTAMP,
    f_str varchar(31),
    f_value_json text,
    f_timestamp timestamp,
    f_timestampz timestamp with time zone,
    f_date_1 date,
    f_time time,
    f_duration varchar(20),
    f_period varchar(20),
    CONSTRAINT Pk_tbl_sample_01 PRIMARY KEY ( id )
);
```

### Versions:

- jOOQ: `3.11.11`
- Java: `java8`
- Database (include vendor): `H2 1.4.199`, `PostgreSQL 10`, `Oracle`
- OS: `any`
- JDBC Driver (include name if inofficial driver):

### Another comment
- This issue only affect when creating table by programmatic by `jooq` api. When generating java class, this issue doesn't happen.
- Just wondering about implementation, why not delegate generate `sql` for field type in `SQLDataType` and it can be overridden in specific `SQLDialect` instead of using switch case in `Tools`

**Comment by katzyn** [comment]
BTW, jOOQ can parse only the valid definitions such as `TIMESTAMP(3) WITH TIME ZONE`, but it translates them into invalid ones with misplaced fractional precision.

**Comment by lukaseder** [comment]
Thank you very much for your report. We'll look into this ASAP

**Comment by knutwannheden** [comment]
Note: `DataType#getCastTypeName()` is also affected by this problem.

**Comment by knutwannheden** [comment]
This bug needs to be fixed in the `DefaultDataType` class. I would like to discuss the implementation options with @lukaseder, before committing to a particular design.

**Comment by lukaseder** [comment]
I agree with @knutwannheden we need some sort of placeholder syntax to be completely dialect agnostic in the generic code of `DefaultDataType`, and without putting existing logic at risk.

I would suggest we follow the plain SQL templating syntax and use curly braces as placeholders. For example, `varchar({length})` would read the length attribute and place it in the template.

Caveats:

- We do not always have length or precision. In that case, we must be careful not to produce the wrong string (e.g. `varchar()`). I would expect that `DefaultDataType::getCastTypeName` could handle this somehow
- We should eagerly pre-calculate the rendering logic by creating a `QueryPart` instead of processing the template at every rendering run. This is actually a more common problem of all plain SQL templating: https://github.com/jOOQ/jOOQ/issues/9298. If this issue can be implemented first, perhaps, we can completely reuse our plain SQL templating functionality for cast expressions.

**Comment by lukaseder** [comment]
This should be fixed via #8029 in jOOQ 3.14.0

**Comment by lukaseder** [comment]
Nope, not yet. #8029 only affects the type name as generated in `CAST` expressions. This is about DDL, whose generation of type expressions could also be improved: #10376.
